### DISCUSSION TITLE
maze_generator: 0.1.0 and beyond...

### DISCUSSION THREAD

**Discussion by mjrlowe** [nestdotland/nest.land#237] [Show and tell]
Hello everyone, I just released [version 0.1.0](https://github.com/mjrlowe/maze_generator/releases/tag/v0.1.0) of [**maze_generator**](https://github.com/mjrlowe/maze_generator), a Javascript module (compatible with Deno, of course) for generating mazes that I've been working on over the last couple of months.

You can use it to generate mazes with 12 different algorithms (I've just added true Prim's and 10print) and there are at least two more algorithms coming soon. 

maze_generator is probably the least useful package on nest.land but I would say it's the most fun! It's also very easy to use. Printing a maze to the console is as simple as `Maze.create().generate().printString()`. 

Here's two examples demonstrating the kinds of images that maze_generator can create:
_[`.display()` currently only works in a HTML document with a canvas element.]_

![10print-60x60](https://user-images.githubusercontent.com/42446683/88220842-74175d80-cc5b-11ea-9412-18f6d798a4b6.png)
_A 60x60 10print maze. `displayMode` is set to `2` (line drawing)._

![colourful maze with no walls](https://user-images.githubusercontent.com/42446683/88219480-90b29600-cc59-11ea-93e4-a3bf380d1e2b.png)
_This one is using the recursive backtracker algorithm, starting in the top left. In the display settings, `displayMode` is set to `1`, `strokeWeight` is set to `0` and `coloringMode` is set to `"distance"`. Essentially you are seeing the cells coloured according to the distance from the top left cell, and walls aren't drawn._

This update also introduces `Maze.createWidget()`, which creates a little interactive HTML widget (inspired by [this](http://jamisbuck.org/mazes/)). This only works in the browser (at least for now) and I still need to do some work on it, but it should be useful for embedding maze generator programs into webpages.

I have a bunch on features planned--including braid mazes, weave mazes, directional bias, seeds and hopefully eventually I will implement mazes of different shapes. I also still need to fully document a couple of features.

Feedback and contributions are welcome: like most people I'm new to Deno and also this is the first module, so I might not have implemented best practices.

I'm enjoying seeing the Deno and nest.land community develop and I look forward to hearing anything you have to say about my hobby project!

**Comment by zorbyte** [comment]
Wowee this looks really neat! :+1:

**Comment by t8** [comment]
This is **so** cool! That colored maze is unlike anything I've ever seen before. What gave you the idea to create something like this, and what is your favorite maze generation algorithm?

**Reply by t8** [reply]
Also, kudos on your documentation over there. It looks very comprehensive and easy to understand.

**Reply by mjrlowe** [reply]
Thanks!

I think the recursive backtracker (randomized depth-first search) algorithm produces the best-looking mazes but some of my other favorites are the Aldous-Broder algorithm for its simplicity and the recursive division algorithm because of how different it is from all the others.

My inspiration mostly is mostly Jamis Buck's stuff on maze algorithms, including [his blog](http://weblog.jamisbuck.org/2011/2/7/maze-generation-algorithm-recap.html), his [CoffeeScript implementations](https://github.com/jamis/csmazes) and his ["Mazes for Programmers" book](http://www.mazesforprogrammers.com/).

Walter D Pullen has has made some great resources on mazes, including the desktop maze app [Daedalus](http://www.astrolog.org/labyrnth/daedalus.htm) and the [maze classification page](http://www.astrolog.org/labyrnth/algrithm.htm) of the *Think Labyrinth* website.

See also:
* [The Wikipedia page](https://en.wikipedia.org/wiki/Maze_generation_algorithm)
* [Coding Train: Maze Generator](https://www.youtube.com/watch?v=HyK_Q5rrcr4) (recursive backtracker)
* [EnslavedInTheScrolls' mazes](https://www.reddit.com/user/enslavedinthescrolls)
* [History of the 10 PRINT algorithm](https://10print.org/)

I mostly just wanted to create this so that I have a single Javascript tool that can do all sorts of maze generation things.

**Reply by t8** [reply]
Interesting... I found myself fascinated in that coding train video you linked. Thanks for sharing more info! I had never thought about the complexities of procedural maze generation until you posted this.

**Comment by littledivy** [comment]
This is super cool! Great work.

**Reply by mjrlowe** [reply]
Thank you!
